File: Problems/Intervals/Insert-Interval/Insert-Interval.cs
using System;
using System.Collections.Generic;

// Using linear search technique - Time: O(n)

public class Solution
{
    public int[][] Insert(int[][] intervals, int[] newInterval)
    {
        List<int[]> ans = new List<int[]>();
        int start = newInterval[0];
        int end = newInterval[1];

        foreach (int[] interval in intervals)
        {
            if (start > end)
            {
                ans.Add(interval);
            }
            else if (interval[1] < start)
            {
                ans.Add(interval);
            }
            else if (interval[0] > end)
            {
                ans.Add(new int[] { start, end });
                start = end + 1;
                ans.Add(interval);
            }
            else
            {
                start = Math.Min(start, interval[0]);
                end = Math.Max(end, interval[1]);
            }
        }

        if (start <= end)
        {
            ans.Add(new int[] { start, end });
        }

        return ans.ToArray();
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[][] intervals = new int[][] {
            new int[] {1, 2},
            new int[] {3, 5},
            new int [] {6, 7},
            new int [] {8, 10},
            new int[] {12, 16}
        };

        int[] newInterval = new int[] { 4, 8 };

        Console.Write("Input: intervals = [");
        foreach (int[] interval in intervals)
        {
            Console.Write("[" + interval[0] + ", " + interval[1] + "]");
            if(interval != intervals[intervals.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.Write("], newInterval = [" + newInterval[0] + ", " + newInterval[1] + "]\n");

        Solution sol = new Solution();
        int[][] ans = sol.Insert(intervals, newInterval);

        Console.Write("Output: [");
        foreach (int[] a in ans)
        {
            Console.Write("[" + a[0] + ", " + a[1] + "]");
            if(a != ans[ans.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.Write("]\n");
    }
}

File: Problems/Intervals/Insert-Interval/Insert-Interval.cpp
#include <iostream>
#include <vector>

// Using linear search technique - Time: O(n)

class Solution {
public:
    std::vector<std::vector<int>> insert(std::vector<std::vector<int>> &intervals, std::vector<int> &newInterval)
    {
        std::vector<std::vector<int>> ans;
        int start = newInterval[0]; 
        int end = newInterval[1]; 
        
        for (auto &intv : intervals)
        {
            if (start > end)
            {
                ans.push_back(intv);
            }
            else if (intv[0] > end)
            {
                ans.push_back({start, end});
                start = end + 1;
                ans.push_back(intv);
            }
            else if (intv[1] < start)
                ans.push_back(intv);
            else
            {
                start = std::min(start, intv[0]);
                end = std::max(end, intv[1]);
            }
        }

        if (start <= end)
        {
            ans.push_back({start, end});
        }
        
        return ans;
    }
};

int main()
{
    std::vector<std::vector<int>> intervals = {{1,2},{3, 5}, {6, 7}, {8, 10}, {12, 16}};
    std::vector<int> newInterval = {4, 8};

    std::cout << "Input: intervals = [";
    for (const auto &interval : intervals)
    {
        std::cout << "[" << interval[0] << "," << interval[1] << "]";
        if (&interval != &intervals.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "], newInterval = [" << newInterval[0] << "," << newInterval[1] << "]" << std::endl;


    Solution sol;
    std::vector<std::vector<int>> ans = sol.insert(intervals, newInterval);

    std::cout << "Output: [";
    for (const auto &a : ans)
    {
        std::cout << "[" << a[0] << "," << a[1] << "]";
        if (&a != &ans.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    return 0;
}